Add search api query

Akronix opened the pull request:
#### Proposed Changes

* Added api query using axios: searchUsers

#### Testing Instructions

* Still missing a test?

Add helper function: SearchUsers to query users by username. Helps #965 and #960.

Use it like this:
```javascript
import { searchUsers } from '@/modules/users/client/utils/search.api';
```

simison commented:
To me `utils` is a bit unexpected — I expected something like "service", "data", "api"? Utils in programming are typically little helper functions.

simison commented:
Let's add the older in the linter list here:
https://github.com/Trustroots/trustroots/blob/ffca3ec877f295dd0140c606e865d1f9c9a7a10a/.eslintrc.js#L121

simison commented on modules/users/client/utils/search.api.js:
```diff
@@ -0,0 +1,5 @@
+import axios from 'axios';
+
+export async function searchUsers(query) {
```
The current filename `search.api.js` doesn't match method name `searchUsers`:

How about we rename the file to `search-users.api.js`?

Akronix replied:
yeah, I thought the same, but I also thought that maybe further methods might be added in the future and they might be living in the same source file, or am I wrong?

simison replied:
Thanks for elaborating on the choice!

>further methods might be added in the future and they might be living in the same source file

If that happens, we can rename the file. Even with additional methods, it'll continue exposing "user search".

Generally speaking, it's better to not name files and methods "in case" but to accurately describe what they do right now. [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and all that. 🙂

simison replied:
Oh I suppose you were looking at the module name and thinking that no need to repeat that?

That would be valid point, but imagine if we'd have multiple "search" files open in the editor, it'll become pretty unwieldly fast.

Akronix replied:
Done.

Akronix commented:
> To me `utils` is a bit unexpected — I expected something like "service", "data", "api"? Utils in programming are typically little helper functions.
> 

Well, I didn't come up with a better name, but aren't they helper functions at the end? in this case, it is a api query, but in the future there could be functions that retrieve data only from the client side or do some javascript computations or wait for user response or whatever.

> Since `services` is colliding with Angular stuff, perhaps `data` or `api-services`?

Following that these names seem too restrictive too me, while `services` was not. But if you people have a better idea for a name, feel free to propose it and rename it.

Akronix commented:
> Let's add the older in the linter list here:
> 
> [trustroots/.eslintrc.js](https://github.com/Trustroots/trustroots/blob/ffca3ec877f295dd0140c606e865d1f9c9a7a10a/.eslintrc.js#L121)
> 
> Line 121 in [ffca3ec](/Trustroots/trustroots/commit/ffca3ec877f295dd0140c606e865d1f9c9a7a10a)
>  'modules/**/client/components/**',

I'm sorry, I didn't get this one :confused:

simison commented:
> aren't they helper functions at the end? 

Kinda everything is. :-) What I mean is that it feels too generic currently and just by looking at the folder name developer couldn't tell what they are going to find there and it easily becomes dumping ground of everything.

>in this case, it is a api query, but in the future there could be functions that retrieve data only from the client side or do some javascript computations or wait for user response or whatever.

Yep, in future, we'll probably have some library handling the data layer. If we use e.g. [Redux](https://redux.js.org/), we'll have "actions", "reducers" and "filters" folders — perhaps all under something called "data"? Anyway, that's future and what we have now are pretty straightforward API services.

What do you think about calling it `data` and seeing how that feels long run? That would accomdoate not only "requesting data from the API" but also "manipulating or filtering data in the app state".

simison commented:
> > Let's add the older in the linter list here:
>  > trustroots/.eslintrc.js
> 
> I'm sorry, I didn't get this one 😕

`.eslintrc.js` contains configurations for our [eslint](https://eslint.org/) linter and new React migration related files are linted differently from the other files. Hence we're explicitly listing them there for now:

https://github.com/Trustroots/trustroots/blob/ffca3ec877f295dd0140c606e865d1f9c9a7a10a/.eslintrc.js#L116-L124

So just adding a new line `'modules/**/client/our_new_folder/**',` would do :thumbsup:

mrkvon commented:
Regarding the folder naming discussion, let me know how it'll get resolved. In #827 I'm using `modules/**/client/api/*.api.js` for keeping functions for getting data from api. Would be good to keep it consistent.
Also /docs/React.md will need an update.

Akronix commented:
I'm waiting for others to express their preferences. Let's leave a couple of days to decide.

guaka commented:
> This stuff doesn't matter to me as long as things are predictable.

+1

On Mon, 17 Dec 2018 at 18:29, mrkvon <notifications@github.com> wrote:

> preferences?
>
> I wouldn't look for api calls in utils directory, so I prefer other names
> (api, data, ?effects). If it will be hard to find a consensus on this, I'm
> comfortable with a little bit of inconsistency until things crystallize. As
> long as we don't reuse the directories with angular stuff in them. This
> stuff doesn't matter to me as long as things are predictable.
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Trustroots/trustroots/pull/984#issuecomment-447928774>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADVyMk3wlhiEEnt-U8Egqtp-zA1FXEsks5u59RggaJpZM4ZVGUL>
> .
>
-- 
sent from my phone

mrkvon commented:
rebased and resolved a conflict

Akronix commented:
I renamed from /utils/ to /api/ since I saw more consensus on that.
@simison the .eslintrc thing was already done before, but I've updated it now accordingly

Akronix commented:
Merge this if you find it OK.

simison commented on package-lock.json:
```diff
@@ -6672,12 +6672,14 @@
         "balanced-match": {
```
Let's remove `package-lock` changes from this PR since we don't modify `package.json` neither: changes seem to be just differences in NPMs of macos/linux.

Akronix replied:
Done. Thank your for the suggestion, I didn't know that one.

simison commented:
Thanks for the path change! 🙌 Much appreciated.

Akronix commented:
Hey folks, I have being testing this function and I find out this:

The function returns an object with this shape:
```
{
  "data": [
    {
      "gender": "",
      "avatarSource": "local",
      "avatarUploaded": true,
      "displayName": "Kareem Gerlach",
      "username": "admin1",
      "emailHash": "56f3ff75b9bdd4668a1b71061f048230",
      "updated": "2018-12-08T23:15:04.841Z",
      "score": 1.1
    }
  ],
  "status": 200,
  "statusText": "OK",
  "headers": {
    "connection": "keep-alive",
    "content-length": "207",
    "content-security-policy-report-only": "default-src 'self'; script-src 'unsafe-eval' 'unsafe-inline' 'self' *.facebook.com *.facebook.net *.fbcdn.net *.twitter.com *.google-analytics.com *.gstatic.com maitreapp.co ajax.googleapis.com 'nonce-13805903-ffe8-463a-9aa1-4a69ce66e3cf' ws://localhost:35729 localhost:35729; font-src 'self' data:; style-src 'self' 'unsafe-inline'; img-src 'self' grafana.trustroots.org *.tiles.mapbox.com api.mapbox.com *.tile.openstreetmap.org *.earthdata.nasa.gov *.facebook.com *.fbcdn.net *.fbsbx.com *.twitter.com *.google-analytics.com *.gstatic.com *.googleusercontent.com *.g.doubleclick.net gravatar.com i0.wp.com i1.wp.com i2.wp.com ucarecdn.com data:; connect-src 'self' maitreapp.co api.mapbox.com fcm.googleapis.com ws://localhost:35729 localhost:35729; object-src 'self'; media-src 'self'; form-action 'self' trustroots.us9.list-manage.com; frame-ancestors 'none'; child-src 'self' *.twitter.com *.facebook.com; base-uri 'self'; report-uri /api/report-csp-violation",
    "content-type": "application/json; charset=utf-8",
    "date": "Thu, 20 Dec 2018 12:07:02 GMT",
    "etag": "W/\"cf-QTbTKAkDlmmvdCDa5WPvVbhrCs4\"",
    "expect-ct": "max-age=30, report-uri=\"http://localhost:3000/api/report-expect-ct-violation\"",
    "strict-transport-security": "max-age=15778476",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-xss-protection": "1; mode=block"
  },
  "config": {
    "transformRequest": {},
    "transformResponse": {},
    "timeout": 0,
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
    "maxContentLength": -1,
    "headers": {
      "Accept": "application/json, text/plain, */*"
    },
    "method": "get",
    "url": "/api/users?search=admin1"
  },
  "request": {}
}

As you can see, only the data attribute is which is actually being expected by the frontend callers.

I think it should be up to the api functions and not to the caller to check the HTTP status returned and return data if was ok or some kind of error otherwise, isn't it?
```

mrkvon commented:
> testing

:+1: :+1: :+1:
Seems like we didn't test it properly before.

Tried to solve this in #1004 now. Not tested, yet.

We can set defaults for axios: https://stackoverflow.com/a/43845580 and one of them is `validateStatus`. It is probably already set to be resolved with codes from 2xx range, otherwise error. We also have the option `transformResponse` which we can use to get only the `data`. I.e. `transformResponse: [({ data }) => data]`.
